src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -746,18 +753,18 @@ public void close(String reason, boolean flushEngine) throws IOException {
                 }
                 changeState(IndexShardState.CLOSED, reason);
             } finally {
-                final Engine engine = this.currentEngineReference.getAndSet(null);
```

Comment:
Engine is closed when the counter goes to 0. IndexShard.close() only decrements it, the closing happens later in closeInternal(). Does that make sense?